### PR TITLE
Partial revert #3731, fix errcheck, add comments

### PR DESCRIPTION
**Pull request by dnr**
**What changed?**
- Revert the changes in #3731 for the scheduler workflow
- Fix errcheck warnings by panicking instead
- Add some comments

**Why?**
The logic in #3731 for errors from Cancel/TerminateWorkflow was wrong, it shouldn't return from a workflow.
The SideEffect errors can't happen by inspection since the types match, but in any case, we can safely panic since it's in a workflow context and the worst case is the workflow gets stuck.

**How did you test it?**
I wrote a unit test that simulated a failure from the CancelWorkflow local activity and tested it before and after. Unfortunately the workflow testing framework seems to be broken for local activity retries, so the test needs to have incorrect logic to match the bug in the framework. So it seems not worth including, though I can if we want.

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/worker/scheduler/workflow.go [hunk 1]
```diff
@@ -908,60 +887,57 @@ func (s *scheduler) startLongPollWatcher(ex *commonpb.WorkflowExecution) {
...
+	// Note: the local activity has completed (or failed) here but we'll still wait until we
+	// observe the workflow close (with a watcher) to start the next one.
+	// If this failed, that's okay, we'll try it again the next time we try to take an action.
 }
 
-func (s *scheduler) newUUIDString() (string, error) {
+func (s *scheduler) newUUIDString() string {
 	if len(s.uuidBatch) == 0 {
-		if err := workflow.SideEffect(s.ctx, func(ctx workflow.Context) interface{} {
+		panicIfErr(workflow.SideEffect(s.ctx, func(ctx workflow.Context) interface{} {
```
Comment (yux0) [inline service/worker/scheduler/workflow.go:926]: To the worst, the worker will stuck in a crash loop?

Comment (dnr) [reply service/worker/scheduler/workflow.go:926]: It's inside a workflow task.. the sdk will catch it and the workflow will be blocked

**Comment by MichaelSnowden** [comment]
> The logic in https://github.com/temporalio/temporal/pull/3731 for errors from Cancel/TerminateWorkflow was wrong, it shouldn't return from a workflow.

Out of curiosity, why can't we return errors like this from a workflow?

**Comment by dnr** [comment]
> Out of curiosity, why can't we return errors like this from a workflow?

It'll end the workflow. It's supposed to be a long-running workflow

File: service/worker/scheduler/workflow.go [hunk 2]
```diff
@@ -690,7 +676,8 @@ func (s *scheduler) addStart(nominalTime, actualTime time.Time, overlapPolicy en
 }
 
 // processBuffer should return true if there might be more work to do right now.
-func (s *scheduler) processBuffer() (bool, error) {
+//nolint:revive
```
Comment (alexshtin) [inline service/worker/scheduler/workflow.go:679]: What does `revive` complain about here?

**Review by alexshtin — APPROVED** [review]

File: service/worker/scheduler/workflow.go (hunk 2 above)
Comment (dnr) [reply service/worker/scheduler/workflow.go:679]: Anyone is welcome to take a try at simplifying this, though

**Pull request merged**